website/content/v1.5/talos-guides/install/bare-metal-platforms/secureboot.md

```diff
@@ -0,0 +1,239 @@
+
+```yaml
+# tpm-disk-encryption.yaml
+machine:
+  systemDiskEncryption:
+    ephemeral:
+      provider: luks2
+      keys:
+        - slot: 0
+          tpm: {}
+    state:
+      provider: luks2
+      keys:
+        - slot: 0
+          tpm: {}
+```
+
+Generate machine configuration:
+
+```shell
+talosctl gen config <cluster-name> https://<endpoint>:6443 --install-image=ghcr.io/<user>/installer-amd64-secureboot:{{< release >}} --install-disk=/dev/sda --config-patch @tpm-disk-encryption.yaml
+```
+
+Apply machine configuration to the node:
+
+```shell
+talosctl -n <IP> apply-config --insecure -f controlplane.yaml
+```
+
+Talos will perform the installation to the disk and reboot the node.
+Please make sure that the ISO image is not attached to the node anymore, otherwise the node will boot from the ISO image again.
+
+Once the node is rebooted, verify that the node is running in secure boot mode:
+
+```shell
+talosctl -n <IP> --talosconfig=talosconfig get securitystate
+```
+
+## Upgrading Talos Linux
+
+Any change to the boot asset (kernel, initramfs, kernel command line) requires the UKI to be regenerated and the installer image to be rebuilt.
+Follow the steps above to generate new installer image updating the boot assets: use new Talos version, add a system extension, or modify the kernel command line.
+Once the new `installer` image is pushed to the registry, [upgrade]({{< relref "../../upgrading-talos" >}}) the node using the new installer image.
+
+It is important to preserve the UKI signing key and the PCR signing key, otherwise the node will not be able to boot with the new UKI and unlock the encrypted partitions.
+
+## Disk Encryption with TPM
+
+When encrypting the disk partition for the first time, Talos Linux generates a random disk encryption key and seals (encrypts) it with the TPM device.
+The TPM unlock policy is configured to trust the policy signed by the PCR signing key.
+This way TPM unlocking doesn't depend on the exact PCR measurements, but rather on the policy signed by the PCR signing key.
+
+When the UKI image is generated, the UKI is measured and expected measurements are combined into TPM unlock policy and signed with the PCR signing key.
+During the boot process, `systemd-stub` component of the UKI performs measurements of the UKI sections into the TPM device.
+Talos Linux during the boot appends to the PCR register the measurements of the boot phases, and once the boot reaches the point of mounting the encrypted disk partition,
+the signed policy from the UKI is used to unlock the TPM, and TPM unseals the disk encryption key which is then used to unlock the disk partition.
+
+During the upgrades, as long as the new UKI is contains PCR policy signed with the same PCR signing key, the disk partition will be unlocked successfully.
+
+Disk encryption is also tied to the state of PCR registry 7, so that it unlocks only if SecureBoot is enabled.
```

Comment:
probably add a note syaing if secureboot keys are added/removed/updated or secureboot disabled the PCR7 state will change

Reply:
oh, I didn't know it also hashes the enrolled keys... so if you enroll more key, disk encryption is disabled? it would be weird, isn't it?

Reply:
got you, updating

Reply:
*updated* is hard to figure if it's existing ones or update the whole db :sweat_smile:

Reply:
yep, it first it sounded weird, but then it made total sense to me ;)